crypto/keys/signer_factory.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keys
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian/crypto/keyspb"
+)
+
+// ProtoHandler uses the information in a protobuf message to obtain a crypto.Signer.
+// For example, the protobuf message may contain a key or identify where a key can be found.
+type ProtoHandler func(context.Context, proto.Message) (crypto.Signer, error)
+
+// ProtoGenerator creates a new private key based on a key specification.
+// It returns a proto that can be passed to a ProtoHandler to get a crypto.Signer.
+type ProtoGenerator func(context.Context, *keyspb.Specification) (proto.Message, error)
+
+// SignerFactory produces a crypto.Signer from a protobuf message describing a key.
+// If SignerFactory.Generate != nil, it can also generate new private keys.
+type SignerFactory struct {
+	// Generate creates a new private key based on a key specification.
+	// It returns a proto that can be passed to NewSigner() to get a crypto.Signer.
+	// If nil, key generation will not be possible.
+	Generate ProtoGenerator
+
+	// handlers convert a protobuf message into a crypto.Signer.
+	handlers map[string]ProtoHandler
```

Comment:
It's what I started with, but...
1. I've known people to be opposed to globals (e.g. @daviddrysdale lamented the scattering of flags through the codebase). 
1. I preferred the explicitness of adding a handler, rather than it being a side-effect of importing a package, since it's quite an important decision - do you really want unencrypted keys in your storage, for instance? It also means it's easier to control which handlers are available to tests.
1. Scoping it to a SignerFactory means you can start servers/signers in-process (like the integration tests do) and not be forced to share the same set of handlers.

Having said that, I do recognise that making the handlers global and having them register on package init is a more common pattern. I can put together a commit that makes that change and see how it compares if you like?